File: polars/expr_str.go
package polars

import (
	pb "github.com/polars-go-bridge/proto"
)

// StrLenBytes 计算字符串字节长度
func (e Expr) StrLenBytes() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrLenBytes{
				StrLenBytes: &pb.StringFunction{Expr: e.inner},
			},
		},
	}
}

// StrLenChars 计算字符串字符长度
func (e Expr) StrLenChars() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrLenChars{
				StrLenChars: &pb.StringFunction{Expr: e.inner},
			},
		},
	}
}

// StrContains 判断字符串是否包含子串/正则
func (e Expr) StrContains(pattern string, literal bool) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrContains{
				StrContains: &pb.StringContains{
					Expr:    e.inner,
					Pattern: pattern,
					Literal: literal,
				},
			},
		},
	}
}

// StrStartsWith 判断字符串是否以指定前缀开头
func (e Expr) StrStartsWith(prefix string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrStartsWith{
				StrStartsWith: &pb.StringStartsWith{
					Expr:   e.inner,
					Prefix: prefix,
				},
			},
		},
	}
}

// StrEndsWith 判断字符串是否以指定后缀结尾
func (e Expr) StrEndsWith(suffix string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrEndsWith{
				StrEndsWith: &pb.StringEndsWith{
					Expr:   e.inner,
					Suffix: suffix,
				},
			},
		},
	}
}

// StrExtract 使用正则提取分组
func (e Expr) StrExtract(pattern string, groupIndex uint32) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrExtract{
				StrExtract: &pb.StringExtract{
					Expr:       e.inner,
					Pattern:    pattern,
					GroupIndex: groupIndex,
				},
			},
		},
	}
}

// StrReplace 替换第一个匹配
func (e Expr) StrReplace(pattern string, value string, literal bool) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrReplace{
				StrReplace: &pb.StringReplace{
					Expr:    e.inner,
					Pattern: pattern,
					Value:   value,
					Literal: literal,
				},
			},
		},
	}
}

// StrReplaceAll 替换全部匹配
func (e Expr) StrReplaceAll(pattern string, value string, literal bool) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrReplaceAll{
				StrReplaceAll: &pb.StringReplace{
					Expr:    e.inner,
					Pattern: pattern,
					Value:   value,
					Literal: literal,
				},
			},
		},
	}
}

// StrToLowercase 转为小写
func (e Expr) StrToLowercase() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrToLowercase{
				StrToLowercase: &pb.StringFunction{Expr: e.inner},
			},
		},
	}
}

// StrToUppercase 转为大写
func (e Expr) StrToUppercase() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrToUppercase{
				StrToUppercase: &pb.StringFunction{Expr: e.inner},
			},
		},
	}
}

// StrStripChars 修剪指定字符（空字符串表示空白字符）
func (e Expr) StrStripChars(chars string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrStripChars{
				StrStripChars: &pb.StringStripChars{
					Expr:  e.inner,
					Chars: chars,
				},
			},
		},
	}
}

// StrSlice 字符串切片（length 可选）
func (e Expr) StrSlice(offset int64, length ...uint64) Expr {
	var lengthPtr *uint64
	if len(length) > 0 {
		l := length[0]
		lengthPtr = &l
	}

	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrSlice{
				StrSlice: &pb.StringSlice{
					Expr:   e.inner,
					Offset: offset,
					Length: lengthPtr,
				},
			},
		},
	}
}

// StrSplit 按分隔符拆分
func (e Expr) StrSplit(by string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrSplit{
				StrSplit: &pb.StringSplit{
					Expr: e.inner,
					By:   by,
				},
			},
		},
	}
}

// StrPadStart 左侧填充
func (e Expr) StrPadStart(length uint64, fillChar string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrPadStart{
				StrPadStart: &pb.StringPad{
					Expr:     e.inner,
					Length:   length,
					FillChar: fillChar,
				},
			},
		},
	}
}

// StrPadEnd 右侧填充
func (e Expr) StrPadEnd(length uint64, fillChar string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrPadEnd{
				StrPadEnd: &pb.StringPad{
					Expr:     e.inner,
					Length:   length,
					FillChar: fillChar,
				},
			},
		},
	}
}

File: rust/src/expr_str.rs
use polars::prelude::*;

use crate::error::BridgeError;
use crate::executor::build_expr;
use crate::proto;

pub fn build_string_expr(kind: &proto::expr::Kind) -> Option<Result<Expr, BridgeError>> {
    use proto::expr::Kind;

    match kind {
        Kind::StrLenBytes(func) => Some(build_unary(func, "StrLenBytes", |expr| {
            expr.str().len_bytes()
        })),
        Kind::StrLenChars(func) => Some(build_unary(func, "StrLenChars", |expr| {
            expr.str().len_chars()
        })),
        Kind::StrContains(contains) => Some(build_contains(contains)),
        Kind::StrStartsWith(starts) => Some(build_starts_with(starts)),
        Kind::StrEndsWith(ends) => Some(build_ends_with(ends)),
        Kind::StrExtract(extract) => Some(build_extract(extract)),
        Kind::StrReplace(replace) => Some(build_replace(replace)),
        Kind::StrReplaceAll(replace) => Some(build_replace_all(replace)),
        Kind::StrToLowercase(func) => Some(build_unary(func, "StrToLowercase", |expr| {
            expr.str().to_lowercase()
        })),
        Kind::StrToUppercase(func) => Some(build_unary(func, "StrToUppercase", |expr| {
            expr.str().to_uppercase()
        })),
        Kind::StrStripChars(strip) => Some(build_strip_chars(strip)),
        Kind::StrSlice(slice) => Some(build_slice(slice)),
        Kind::StrSplit(split) => Some(build_split(split)),
        Kind::StrPadStart(pad) => Some(build_pad_start(pad)),
        Kind::StrPadEnd(pad) => Some(build_pad_end(pad)),
        _ => None,
    }
}

fn build_unary<F>(
    func: &proto::StringFunction,
    name: &str,
    op: F,
) -> Result<Expr, BridgeError>
where
    F: FnOnce(Expr) -> Expr,
{
    let expr = build_inner_expr(&func.expr, name)?;
    Ok(op(expr))
}

fn build_contains(contains: &proto::StringContains) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&contains.expr, "StrContains")?;
    let pattern = lit(contains.pattern.as_str());

    if contains.literal {
        Ok(expr.str().contains_literal(pattern))
    } else {
        Ok(expr.str().contains(pattern, true))
    }
}

fn build_starts_with(starts: &proto::StringStartsWith) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&starts.expr, "StrStartsWith")?;
    Ok(expr.str().starts_with(lit(starts.prefix.as_str())))
}

fn build_ends_with(ends: &proto::StringEndsWith) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&ends.expr, "StrEndsWith")?;
    Ok(expr.str().ends_with(lit(ends.suffix.as_str())))
}

fn build_extract(extract: &proto::StringExtract) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&extract.expr, "StrExtract")?;
    Ok(expr
        .str()
        .extract(lit(extract.pattern.as_str()), extract.group_index as usize))
}

fn build_replace(replace: &proto::StringReplace) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&replace.expr, "StrReplace")?;
    Ok(expr.str().replace(
        lit(replace.pattern.as_str()),
        lit(replace.value.as_str()),
        replace.literal,
    ))
}

fn build_replace_all(replace: &proto::StringReplace) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&replace.expr, "StrReplaceAll")?;
    Ok(expr.str().replace_all(
        lit(replace.pattern.as_str()),
        lit(replace.value.as_str()),
        replace.literal,
    ))
}

fn build_strip_chars(strip: &proto::StringStripChars) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&strip.expr, "StrStripChars")?;
    let matches_expr = if strip.chars.is_empty() {
        lit(NULL).cast(DataType::String)
    } else {
        lit(strip.chars.as_str())
    };

    Ok(expr.str().strip_chars(matches_expr))
}

fn build_slice(slice: &proto::StringSlice) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&slice.expr, "StrSlice")?;
    let offset = lit(slice.offset);
    let length = match slice.length {
        Some(length) => lit(length),
        None => lit(NULL).cast(DataType::UInt64),
    };

    Ok(expr.str().slice(offset, length))
}

fn build_split(split: &proto::StringSplit) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&split.expr, "StrSplit")?;
    Ok(expr.str().split(lit(split.by.as_str())))
}

fn build_pad_start(pad: &proto::StringPad) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&pad.expr, "StrPadStart")?;
    let fill_char = parse_fill_char(&pad.fill_char, "StrPadStart")?;
    Ok(expr.str().pad_start(lit(pad.length), fill_char))
}

fn build_pad_end(pad: &proto::StringPad) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&pad.expr, "StrPadEnd")?;
    let fill_char = parse_fill_char(&pad.fill_char, "StrPadEnd")?;
    Ok(expr.str().pad_end(lit(pad.length), fill_char))
}

fn build_inner_expr(
    expr: &Option<Box<proto::Expr>>,
    name: &str,
) -> Result<Expr, BridgeError> {
    let expr = expr
        .as_ref()
        .ok_or_else(|| BridgeError::PlanSemantic(format!("{name} has no expr")))?;
    build_expr(expr.as_ref())
}

fn parse_fill_char(value: &str, name: &str) -> Result<char, BridgeError> {
    let mut chars = value.chars();
    let fill_char = chars.next().ok_or_else(|| {
        BridgeError::InvalidArgument(format!("{name} fill_char cannot be empty"))
    })?;

    if chars.next().is_some() {
        return Err(BridgeError::InvalidArgument(format!(
            "{name} fill_char must be a single character"
        )));
    }

    Ok(fill_char)
}
